File: labs/M03-prompts/solution/message_roles.js
/**
 * M03 Lab — Message Roles (Node.js Solution)
 * =============================================
 * Explore how system prompts, user messages, and assistant prefill
 * change Claude's responses to the same question.
 */

import Anthropic from "@anthropic-ai/sdk";
import "dotenv/config";

const client = new Anthropic();
const MODEL = "claude-sonnet-4-20250514";

/**
 * Send a simple user message with no system prompt.
 * @param {string} userMessage
 * @returns {Promise<string>}
 */
async function basicCall(userMessage) {
  const response = await client.messages.create({
    model: MODEL,
    max_tokens: 1024,
    messages: [{ role: "user", content: userMessage }],
  });
  return response.content[0].text;
}

/**
 * Send a user message with a system prompt that sets Claude's role.
 * @param {string} system
 * @param {string} userMessage
 * @returns {Promise<string>}
 */
async function withSystemPrompt(system, userMessage) {
  const response = await client.messages.create({
    model: MODEL,
    max_tokens: 1024,
    system,
    messages: [{ role: "user", content: userMessage }],
  });
  return response.content[0].text;
}

/**
 * Use assistant prefill to guide the response format.
 * @param {string} system
 * @param {string} userMessage
 * @param {string} assistantPrefill
 * @returns {Promise<string>}
 */
async function withPrefill(system, userMessage, assistantPrefill) {
  const response = await client.messages.create({
    model: MODEL,
    max_tokens: 1024,
    system,
    messages: [
      { role: "user", content: userMessage },
      { role: "assistant", content: assistantPrefill },
    ],
  });
  return assistantPrefill + response.content[0].text;
}

// ─── Main ───────────────────────────────────────────────────────────────────

const question = "What is a UCC-1 filing?";

const systemPrompt =
  "You are a UCC filing expert. You specialize in Uniform Commercial Code " +
  "filings, secured transactions, and lien searches. Provide accurate, " +
  "concise explanations using proper legal terminology while remaining " +
  "accessible to non-lawyers.";

const prefill = "## Analysis\n";

// --- Call 1: Basic (no system prompt) ---
console.log("=".repeat(60));
console.log("CALL 1: Basic — No System Prompt");
console.log("=".repeat(60));
try {
  const result = await basicCall(question);
  console.log(result);
} catch (e) {
  console.log(`[ERROR] ${e.message}`);
}

// --- Call 2: With system prompt ---
console.log("\n" + "=".repeat(60));
console.log("CALL 2: With System Prompt (UCC Expert)");
console.log("=".repeat(60));
try {
  const result = await withSystemPrompt(systemPrompt, question);
  console.log(result);
} catch (e) {
  console.log(`[ERROR] ${e.message}`);
}

// --- Call 3: With system prompt + prefill ---
console.log("\n" + "=".repeat(60));
console.log("CALL 3: With System Prompt + Assistant Prefill");
console.log("=".repeat(60));
try {
  const result = await withPrefill(systemPrompt, question, prefill);
  console.log(result);
} catch (e) {
  console.log(`[ERROR] ${e.message}`);
}

File: labs/M03-prompts/solution/message_roles.py
"""
M03 Lab — Message Roles (Solution)
====================================
Explore how system prompts, user messages, and assistant prefill
change Claude's responses to the same question.
"""

import anthropic
from dotenv import load_dotenv

load_dotenv()

client = anthropic.Anthropic()
MODEL = "claude-sonnet-4-6"

# Assistant prefill is not universal. claude-sonnet-4-6 rejects it outright --
# "This model does not support assistant message prefill" -- so Call 3 below
# uses a model that does. That is the lesson as much as the technique is: a
# prompting trick you rely on can be unavailable on the model you deploy, so
# check before you build on it.
PREFILL_MODEL = "claude-haiku-4-5-20251001"


def basic_call(user_message: str) -> str:
    """Send a simple user message with no system prompt."""
    response = client.messages.create(
        model=MODEL,
        max_tokens=1024,
        messages=[{"role": "user", "content": user_message}],
    )
    return response.content[0].text


def with_system_prompt(system: str, user_message: str) -> str:
    """Send a user message with a system prompt that sets Claude's role."""
    response = client.messages.create(
        model=MODEL,
        max_tokens=1024,
        system=system,
        messages=[{"role": "user", "content": user_message}],
    )
    return response.content[0].text


def with_prefill(system: str, user_message: str, assistant_prefill: str) -> str:
    """Use assistant prefill to guide the response format."""
    response = client.messages.create(
        model=PREFILL_MODEL,
        max_tokens=1024,
        system=system,
        messages=[
            {"role": "user", "content": user_message},
            # rstrip(): the API rejects a final assistant message ending in
            # whitespace ("final assistant content cannot end with trailing
            # whitespace"), and the prefill below ends in a newline. The
            # un-stripped text is still used for the value returned, so the
            # formatting the prefill is teaching is preserved.
            {"role": "assistant", "content": assistant_prefill.rstrip()},
        ],
    )
    return assistant_prefill + response.content[0].text


# ─── Main ───────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    question = "What is a UCC-1 filing?"

    system_prompt = (
        "You are a UCC filing expert. You specialize in Uniform Commercial Code "
        "filings, secured transactions, and lien searches. Provide accurate, "
        "concise explanations using proper legal terminology while remaining "
        "accessible to non-lawyers."
    )

    prefill = "## Analysis\n"

    # --- Call 1: Basic (no system prompt) ---
    print("=" * 60)
    print("CALL 1: Basic — No System Prompt")
    print("=" * 60)
    try:
        result = basic_call(question)
        print(result)
    except Exception as e:
        print(f"[ERROR] {e}")

    # --- Call 2: With system prompt ---
    print("\n" + "=" * 60)
    print("CALL 2: With System Prompt (UCC Expert)")
    print("=" * 60)
    try:
        result = with_system_prompt(system_prompt, question)
        print(result)
    except Exception as e:
        print(f"[ERROR] {e}")

    # --- Call 3: With system prompt + prefill ---
    print("\n" + "=" * 60)
    print("CALL 3: With System Prompt + Assistant Prefill")
    print("=" * 60)
    try:
        result = with_prefill(system_prompt, question, prefill)
        print(result)
    except Exception as e:
        print(f"[ERROR] {e}")
